edb/server/protocol/auth_ext/local.py

```diff
@@ -171,3 +186,100 @@ async def authenticate(self, db: Any, input: dict[str, Any]):
             )
 
         return local_identity
+
+    async def get_identity_and_secret(self, db: Any, input: dict[str, Any]):
+        if 'handle' not in input:
+            raise errors.InvalidData("Missing 'handle' in data")
+
+        handle = input["handle"]
+        r = await execute.parse_execute_json(
+            db=db,
+            query="""
+with
+  handle := <str>$handle,
+select ext::auth::PasswordCredential {
+  password_hash,
+  identity: { * }
+} filter .identity.handle = handle""",
+            variables={
+                "handle": handle,
+            }
+        )
+
+        result_json = json.loads(r.decode())
+        if len(result_json) != 1:
+            raise errors.NoIdentityFound()
+        password_cred = result_json[0]
+
+        local_identity = data.LocalIdentity(
+            **password_cred["identity"]
+        )
+        secret = base64.b64encode(
+            hashlib.sha256(password_cred['password_hash'].encode()).digest()
+        ).decode()
+
+        return (local_identity, secret)
+
+    async def validate_reset_secret(
+        self, db: Any, identity_id: str, secret: str
+    ):
+
+        r = await execute.parse_execute_json(
+            db=db,
+            query="""\
+with
+  identity_id := <uuid><str>$identity_id,
+select ext::auth::PasswordCredential { password_hash, identity: { * } }
+filter .identity.id = identity_id;""",
+            variables={
+                "identity_id": identity_id,
+            },
+        )
+
+        result_json = json.loads(r.decode())
+        if len(result_json) != 1:
+            raise errors.NoIdentityFound()
+        password_cred = result_json[0]
+
+        local_identity = data.LocalIdentity(
+            **password_cred["identity"]
+        )
+
+        current_secret = base64.b64encode(
+            hashlib.sha256(password_cred['password_hash'].encode()).digest()
+        ).decode()
+
+        return local_identity if secret == current_secret else None
+
+    async def update_password(
+        self, db: Any, identity_id: str, secret: str, input: dict[str, Any]
+    ):
+        if 'password' not in input:
+            raise errors.InvalidData("Missing 'password' in data")
+
+        password = input["password"]
+
+        local_identity = await self.validate_reset_secret(
+            db, identity_id, secret)
+
+        if local_identity is None:
+            raise errors.InvalidData("Invalid 'reset_token'")
+
+        # TODO: check if race between validating secret and updating password
+        #       is a problem
```

Comment:
Yeah, it would be nice to take a row-lock on the `PasswordCredential` when validating. We don't have a mechanism for that in EdgeQL yet though, right?